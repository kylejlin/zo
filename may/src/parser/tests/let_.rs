use super::*;

#[test]
fn let_() {
    let src = r#"
let three = succ(two)

add(three, three)
"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(cst);
}

#[test]
fn let_val_is_let() {
    let src = r#"
let three =
    let two = succ(one)
    succ(two)

add(three, three)
"#;
    let tokens = lex(src).unwrap();
    let cst = parse(tokens).unwrap();
    insta::assert_debug_snapshot!(cst);
}
