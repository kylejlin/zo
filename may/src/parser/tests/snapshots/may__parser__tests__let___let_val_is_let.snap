---
source: may/src/parser/tests/let_.rs
expression: cst
---
Let(
    Let {
        letkw: ByteIndex(
            1,
        ),
        name: Ident {
            value: "three",
            start: ByteIndex(
                5,
            ),
        },
        val: Let(
            Let {
                letkw: ByteIndex(
                    17,
                ),
                name: Ident {
                    value: "two",
                    start: ByteIndex(
                        21,
                    ),
                },
                val: VarOrApp(
                    App(
                        App {
                            callee: Var(
                                Ident {
                                    value: "succ",
                                    start: ByteIndex(
                                        27,
                                    ),
                                },
                            ),
                            lparen: ByteIndex(
                                31,
                            ),
                            args: One(
                                VarOrApp(
                                    Var(
                                        Ident {
                                            value: "one",
                                            start: ByteIndex(
                                                32,
                                            ),
                                        },
                                    ),
                                ),
                            ),
                            rparen: ByteIndex(
                                35,
                            ),
                        },
                    ),
                ),
                next_val: VarOrApp(
                    App(
                        App {
                            callee: Var(
                                Ident {
                                    value: "succ",
                                    start: ByteIndex(
                                        41,
                                    ),
                                },
                            ),
                            lparen: ByteIndex(
                                45,
                            ),
                            args: One(
                                VarOrApp(
                                    Var(
                                        Ident {
                                            value: "two",
                                            start: ByteIndex(
                                                46,
                                            ),
                                        },
                                    ),
                                ),
                            ),
                            rparen: ByteIndex(
                                49,
                            ),
                        },
                    ),
                ),
            },
        ),
        next_val: VarOrApp(
            App(
                App {
                    callee: Var(
                        Ident {
                            value: "add",
                            start: ByteIndex(
                                52,
                            ),
                        },
                    ),
                    lparen: ByteIndex(
                        55,
                    ),
                    args: Snoc(
                        One(
                            VarOrApp(
                                Var(
                                    Ident {
                                        value: "three",
                                        start: ByteIndex(
                                            56,
                                        ),
                                    },
                                ),
                            ),
                        ),
                        VarOrApp(
                            Var(
                                Ident {
                                    value: "three",
                                    start: ByteIndex(
                                        63,
                                    ),
                                },
                            ),
                        ),
                    ),
                    rparen: ByteIndex(
                        68,
                    ),
                },
            ),
        ),
    },
)
